iio: adc: ad9361: re-organize structs to make them opaque to other files

commodo opened the pull request:
This changeset intends to hide away some of the information in `ad9361.h` into a `ad9361_private.h` file which should be available only to `ad9361.c`.

Currently, the only other file related to this driver is `ad9361_conv.c`, but a new driver sub-module will be added in a subsequent series.
This split/re-organization is not perfect, but it is a step into modularizing the driver a bit.

Also, with this set, a new `struct ad9361_rf_phy_state` is split from the `struct ad9361_rf_phy`, and `struct ad9361_phy_platform_data`  is made completely unavailable to `ad9361_conv.c`.
This should only contain state information that is available only to the `ad9361.c` logic, and accessible only via accessor functions.

Hopefully the changes are simple enough [some even a bit big] to be follow-able at review.
Tested on a SidekiqZ2.

Signed-off-by: Alexandru Ardelean <alexandru.ardelean@analog.com>

commodo commented:
Changelog v1 -> v2:
* added entry for SidekiqZ2 to .travis.yml

commodo commented:
oh ; `checkpatch.pl`
will fix tomorrow ;

mhennerich commented on drivers/iio/adc/ad9361_conv.c:
```diff
@@ -639,11 +641,11 @@ int ad9361_dig_tune(struct ad9361_rf_phy *phy, unsigned long max_freq,
-		ad9361_bist_loopback(phy, loopback);
-		ad9361_spi_write(phy->spi, REG_BIST_CONFIG, bist);
+		ad9361_bist_loopback(phy, data.bist_loopback_mode);
+		ad9361_spi_write(phy->spi, REG_BIST_CONFIG, data.bist);
```
Here is a race condition. We cache st->bist_config which is set in either ad9361_bist_tone or ad9361_bist_prbs.
Then we use  ad9361_bist_prbs in here. However when we restore REG_BIST_CONFIG st->bist_config still points to BIST_INJ_RX. So we need an accessor function for  REG_BIST_CONFIG which always updates st->bist_config.

You see this issue when you enable the FIR filter - bist PRBS doesn't get reset ...

commodo commented on drivers/iio/adc/ad9361.c:
```diff
@@ -4170,7 +4170,26 @@ int ad9361_set_ensm_mode(struct ad9361_rf_phy *phy, bool fdd, bool pinctrl)
+{
+	/* FIXME: see if this condition is correct;
+	   looks like it should not be negated, since `bool fdd`*/
+	if (!phy->pdata->fdd)
```
@mhennerich @larsclausen 
[curios] should this condition be inverted ?
i.e.
```
if (phy->pdata->fdd)
   return ad9361_set_ensm_mode(phy, true, false);
```

reason for asking is `int ad9361_set_ensm_mode(struct ad9361_rf_phy *phy, bool fdd, bool pinctrl)`
what happens now is that:
```
if (!phy->pdata->fdd)
   return ad9361_set_ensm_mode(phy, true, false);
```
basically if FDD is false [on pdata] then we call this function with FDD true
it could be correct as-is, but the logic is a bit confusing/non-intuitive;

mhennerich replied:
That's exactly the intended function. You can't do the digital tuning in TDD mode.
Since you require both RX and TX to be operational for it.

commodo commented:
> This patchset is pretty invasive. We need to make sure we sync the no-OS driver the same way since we want to avoid diverging the code bases. The No-OS driver in fact can benefit much more from the header guard, since lot's of people calling into the ad9361 driver directly.
Which causes all kind of undesired behavior.

ack;
maybe we could split the PR into smaller PRs and handle things on a case-by-case;
i wanted to first submit it as a whole, to make it clearer what my intent is with this changeset ;

if there is a concern about this, we can drop it;
but i would prefer to have it, to mitigate future additions to the driver

commodo commented on drivers/iio/adc/ad9361.c:
```diff
@@ -4675,38 +4704,40 @@ static int ad9361_rssi_gain_step_calib(struct ad9361_rf_phy *phy)
-	memset(&phy->fastlock, 0, sizeof(phy->fastlock));
+	struct ad9361_rf_phy_state *st = phy->state;
+
+	st->current_table = -1;
```
ack

commodo commented:
regarding the no-OS changes ; i will work with Dragos to see about integrating things there as well in a similar manner 
[ if this changeset is acceptable ]

commodo commented:
Changelog v2 -> v3:
* added whitespace fixes reported by checkpatch.pl script
* added commit `iio: adc: ad9361: consolidate state struct init/clear`
* added commit ` iio: adc: ad9361: add ad9361_write_bist_reg() accessor to cache bistreg value`

commodo commented:
Changes v3 -> v4:
* rebased on top of new master ( 06747c33fd486c454b3d62a2fb053f1322f101a7 )
  * commit  `iio: adc: ad9631: use ad9361_bist_prbs() instead of ad9361_spi_write()` is now in master via PR https://github.com/analogdevicesinc/linux/pull/64
  * reworked patches to new master

commodo commented:
Note: for some reason, commits on Github are re-ordered vs how they should be.

commodo commented:
Changes v4 -> v5: 
* rebased on top of new master ( 6cfc28c98ae7dd82fdbe61e2b83a0260b6c32679 )
  * commit  `iio: adc: ad9361: Add option to set Reg Gain Step Config2 0x106 [D6:D4]
` is now in master via PR https://github.com/analogdevicesinc/linux/pull/74
  * reworked patches to new master

commodo commented:
Changes v5 -> v6: 
* rebased on top of new master ( ca551e550bd1c53077680610632a04798f1147eb )
  * commit  `.travis.yml: add ZynqMP to the build run` is now in master via PR https://github.com/analogdevicesinc/linux/pull/72
  * reworked patches to new master

commodo commented:
Changes v6 -> v7: 
* rebased on top of new master ( ca1e1c75010da9b5b2158534426cb49ed82ad3b2 ) ; resolved conflict in ad9361.h

commodo commented:
Changes v7 -> v8:
* rebased on top of new master ( a8a90e6f2652dd2544e45855db15f3c1f4668ff3 ) ; resolved conflict in ad9361.h

commodo commented:
if no objections, i'll merge this tomorrow

mhennerich commented:
sounds good
